statistics/handle/handle.go

```diff
@@ -2052,3 +2014,73 @@ func (h *Handle) recordHistoricalStatsMeta(tableID int64, version uint64) error
 	}
 	return nil
 }
+
+// RecordHistoricalStatsToStorage records the given table's stats data to mysql.stats_history
+func (h *Handle) RecordHistoricalStatsToStorage(dbName string, tableInfo *model.TableInfo) (uint64, error) {
+	ctx := context.Background()
+	js, err := h.DumpStatsToJSON(dbName, tableInfo, nil)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	version := uint64(0)
+	for _, value := range js.Columns {
+		version = uint64(*value.StatsVer)
+		if version != 0 {
+			break
+		}
+	}
+	blocks, err := JSONTableToBlocks(js, maxColumnSize)
+	if err != nil {
+		return version, errors.Trace(err)
+	}
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+	_, err = exec.ExecuteInternal(ctx, "begin pessimistic")
+	if err != nil {
+		return version, errors.Trace(err)
+	}
+	defer func() {
+		err = finishTransaction(ctx, exec, err)
+	}()
+	ts := time.Now().Format("2006-01-02 15:04:05.999999")
+
+	const sql = "INSERT INTO mysql.stats_history(table_id, stats_data, seq_no, version, create_time) VALUES (%?, %?, %?, %?, %?)"
+	for i := 0; i < len(blocks); i++ {
+		if _, err := exec.ExecuteInternal(ctx, sql, tableInfo.ID, blocks[i], i, version, ts); err != nil {
+			return version, errors.Trace(err)
+		}
+	}
+	return version, nil
+}
+
+// RecordAnalyzeJob inserts analyze job into mysql.analyze_jobs and gets job ID for further updating job.
+func (h *Handle) RecordAnalyzeJob(job *statistics.AnalyzeJob, procID uint64) error {
+	serverInfo, err := infosync.GetServerInfo()
+	if err != nil {
+		return err
+	}
+	address := fmt.Sprintf("%s:%d", serverInfo.IP, serverInfo.Port)
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	exec := h.mu.ctx.(sqlexec.RestrictedSQLExecutor)
+	ctx := context.TODO()
+	const insertJob = "INSERT INTO mysql.analyze_jobs (table_schema, table_name, partition_name, job_info, state, instance, process_id) VALUES (%?, %?, %?, %?, %?, %?, %?)"
+	_, _, err = exec.ExecRestrictedSQL(ctx, []sqlexec.OptionFuncAlias{sqlexec.ExecOptionUseCurSession}, insertJob, job.DBName, job.TableName, job.PartitionName, job.JobInfo, statistics.AnalyzePending, address, procID)
+	if err != nil {
+		return err
+	}
+	const getJobID = "SELECT LAST_INSERT_ID()"
```

Comment:
Is this possible to directly use `func (s *session) LastInsertID() uint64`?

Reply:
I am not sure about it. We use `session.GetSessionVars().StmtCtx.PrevLastInsertID` to evaluate `LAST_INSERT_ID()`.
https://github.com/pingcap/tidb/blob/7c69e74bd9dad30dc9d724dc9d0b3fc759973b40/expression/builtin_info.go#L358-L363
In `ResetContextOfStmt`, we set `StmtCtx.PrevLastInsertID` rather than `StmtCtx.LastInsertID`.
https://github.com/pingcap/tidb/blob/7c69e74bd9dad30dc9d724dc9d0b3fc759973b40/executor/executor.go#L1904-L1908
IMO, `(*session).LastInsertID` seems to be for the current statement rather than the previous statement.